cumulus_etl/fhir/fhir_schemas.py

```diff
@@ -3,78 +3,123 @@
 from collections import namedtuple
 from functools import partial
 
+import pyarrow
 from fhirclient.models import fhirabstractbase, fhirdate, fhirelementfactory
-import pyspark
 
 
 FhirProperty = namedtuple("FhirProperty", ["name", "json_name", "pytype", "is_list", "of_many", "required"])
 
 
-def create_spark_schema_for_resource(resource_type: str) -> pyspark.sql.types.StructType:
+def pyarrow_schema_from_resource_batch(resource_type: str, batch: list[dict]) -> pyarrow.Schema:
     """
-    Creates a Pyspark StructType schema based off the named resource (like 'Observation').
+    Creates a PyArrow schema based off the named resource (like 'Observation') and batch contents.
 
     Note that this schema will not be deep (fully nested all the way down),
     it will simply be wide (covering each toplevel field, each likely nullable).
+    But it *will* at least include every field contained in the batch.
 
     The primary goal here is to simplify complexity in the consuming SQL so that it
     can assume each column is at least defined.
     """
+    # Examine batch to see the full shape of it, in order to detect any deeply nested fields that we want to make sure
+    # to include in the final schema (normally, we go wide but only as deep as we need to)
+    batch_shape = merge_shape_of_dict(None, batch)
+
+    return create_pyarrow_schema_for_resource(resource_type, batch_shape)
+
+
+def merge_shape_of_dict(total: dict | None, item) -> dict:
```

Comment:
i'm bouncing off this name a bit - perhaps `get_shape_of_merged_dict`?

does this function need a docstring?

Reply:
Yeah OK for sure a docstring and I'll try to think of a better name - my brain said `merge` because it's a recursive thing that is always merging back changes into the master dict. But that can be clearer / better worded. And yeah a docstring. Good call out.

Reply:
OK done - hopefully this is a little clearer (and I included an example in the docstring)